paths_get.py

```diff
@@ -0,0 +1,168 @@
+from typing import Iterable, Tuple, Union, Literal, Sequence, Dict
+
+import numpy as np
+import pandas as pd
+from anndata import AnnData
+from anndata._core.raw import Raw
+
+
+# TODO: allow sequences: ("obsm", "X_pca", [0, 1])
+AnnDataPath = Union[
+    Tuple[Literal["X"], str],  # TODO: Support this or do ("layer", None, str)?
+    Tuple[Literal["layer"], str, str],
+    Tuple[Literal["obs"], str],
+    Tuple[Literal["obsm"], str, Union[int, str]],
+    Tuple[Literal["obsp"], str, Literal[0, 1]],  # TODO: actually needs one more parameter (cell name)
+    Tuple[Literal["raw"], Literal["X"], str],  # TODO: or "raw/X" and "raw/obs"?
+    Tuple[Literal["raw"], Literal["obs"], str],
+]
+
+
+def obs_df(
+    adata: AnnData,
+    paths: Iterable[Union[str, AnnDataPath]] = (),
+    *,
+    gene_symbols: AnnDataPath = None,
+    # no use_raw, no layer.
+) -> pd.DataFrame:
+    paths = list(map(normalize_adpath, paths))
+    names = paths_to_names(paths)
+    columns = {n: get_obs_vector(adata, p, gene_symbols) for n, p in names.items()}
+    return pd.DataFrame(columns, adata.obs_names)
+
+
+def get_obs_vector(adata: Union[AnnData, Raw], path: Union[str, AnnDataPath], gene_symbols: Union[str, AnnDataPath]):
+    path = normalize_adpath(path)
+    gene_symbols = None if gene_symbols is None else normalize_adpath(gene_symbols)
+    
+    if path[0] == "layer":  # X is here after normalizing
+        layer = adata.X if path[1] == "X" else adata.layers[path[1]]
+        idx = adata.var_names if gene_symbols is None else adata.var[gene_symbols]
+        # TODO: can layers also be data frames and have to be handled as obsm?
+        return layer[:, idx == path[2]]
```

Comment:
Please don't have this return a sparse matrix. Just grab the current `sc.get.obs_df` logic for dealing with this bit.

Reply:
sure, that all was just a quick hack!